sandbox/rocky/tf/algos/batch_polopt.py

```diff
@@ -91,6 +91,7 @@ def start_worker(self):
 
     def shutdown_worker(self):
         self.sampler.shutdown_worker()
+        plotter.shutdown_worker()
```

Comment:
isn't this called by atexit? Our API contract is that the algorithm just needs to call init_plot()

Reply:
The issue is when I raise a Keyboardinterrupt , for some reason the interrupt is passed to shutdown worker very late or not passed at all. Till that time the thread continues to do something and then a runtime exception occurs. Once the thread fails then atexit function is called. Here is the traceback: 

Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/home/ray/rllabs/rllab/sandbox/rocky/tf/plotter/plotter.py", line 51, in run
    rollout(env, policy, max_path_length=max_length, animated=True, speedup=5)
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 3613, in get_controller
    yield default
  File "/home/ray/rllabs/rllab/sandbox/rocky/tf/plotter/plotter.py", line 51, in run
    rollout(env, policy, max_path_length=max_length, animated=True, speedup=5)
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 3613, in get_controller
    yield default
  File "/home/ray/rllabs/rllab/sandbox/rocky/tf/plotter/plotter.py", line 51, in run
    rollout(env, policy, max_path_length=max_length, animated=True, speedup=5)
  File "/home/ray/rllabs/rllab/rllab/sampler/utils.py", line 19, in rollout
    a, agent_info = agent.get_action(o)
  File "/home/ray/rllabs/rllab/sandbox/rocky/tf/policies/gaussian_mlp_policy.py", line 161, in get_action
    mean, log_std = [x[0] for x in self._f_dist([flat_obs])]
  File "/home/ray/rllabs/rllab/sandbox/rocky/tf/misc/tensor_utils.py", line 8, in run
    return sess.run(outputs, feed_dict=dict(list(zip(inputs, input_vals))))
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 767, in run
    run_metadata_ptr)
  File "/home/ray/anaconda2/envs/rllab3/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 903, in _run
    raise RuntimeError('Attempted to use a closed Session.')
RuntimeError: Attempted to use a closed Session.

So to prevent this, I explicitly called shutdown thread to stop the thread.

Reply:
That sounds like the plotter API is leaking again.

If the user has to call shutdown_worker(), they should also have to call init_worker(), i.e. they should have to manage the entire lifecycle themselves or not at all. A mix of automatic and manual management makes it very difficult for users to predict the behavior and expectations of the API.

Reply:
Hi,
I tried to avoid using explicit call to shutdown. Here is the current status:
1. When we generate a keyboard interrupt, threading module explicitly calls to t.join before the call to atexit function. To overcome this , I passed the thread as daemon which avoids the call to join method.
2. When we are at the end of the iteration, session is closed in main thread but there are still some samples in queue. Hence the thread tries to use rollout for plotting them. As a result we observe this exception "RuntimeError: Attempted to use a closed Session".
To overcome this issue, I used a function call sess._closed to check if it is closed or not but still rollout fails with the same exception. I am inspecting that rollout takes some time to plot the value. By the time rollout is executed , session is closed.
To overcome this issue, I am planning to remove the close session call from the main process and call it in thread so that session value exists when plots are performed. Let me know your views on this implementation.

Reply:
Managing the session lifecycle between two threads is going to be very difficult, as you have found.

Can the plotter thread use its own session?